eclipse-collections-code-generator/src/main/resources/impl/collection/mutable/abstractMultiReaderPrimitiveCollection.stg

```diff
@@ -0,0 +1,554 @@
+import org.eclipse.collections.api.bag.primitive.Mutable<name>Bag;
+import org.eclipse.collections.api.block.function.primitive.<name>ToObjectFunction;
+import org.eclipse.collections.api.block.function.primitive.Object<name>ToObjectFunction;
+import org.eclipse.collections.api.block.predicate.primitive.<name>Predicate;
+import org.eclipse.collections.api.block.procedure.primitive.<name>Procedure;
+import org.eclipse.collections.api.collection.MutableCollection;
+import org.eclipse.collections.api.collection.primitive.Immutable<name>Collection;
+import org.eclipse.collections.api.collection.primitive.Mutable<name>Collection;
+import org.eclipse.collections.api.iterator.Mutable<name>Iterator;
+import org.eclipse.collections.api.iterator.<name>Iterator;
+import org.eclipse.collections.api.list.primitive.Mutable<name>List;
+import org.eclipse.collections.api.set.primitive.Mutable<name>Set;
+import org.eclipse.collections.impl.lazy.primitive.Lazy<name>IterableAdapter;
+import net.jcip.annotations.ThreadSafe;
+
+/**
+ * This file was automatically generated from template file abstractMultiReaderPrimitiveCollection.stg.
+ *
+ * @since 8.1.
+ */
+@ThreadSafe
+public abstract class AbstractMultiReader<name>Collection
+        implements Mutable<name>Collection, Serializable
+{
+    private static final long serialVersionUID = 1L;
+
+    private final ReadWriteLock lock;
+
+    private final Mutable<name>Collection collection;
+
+    protected AbstractMultiReader<name>Collection(Mutable<name>Collection collection)
+    {
+        this(collection, null);
+    }
+
+    protected AbstractMultiReader<name>Collection(Mutable<name>Collection collection, ReadWriteLock newLock)
+    {
+        if(collection == null)
+        {
+            throw new IllegalArgumentException("Cannot create a AbstractMultiReader<name>Collection on a null collection");
+        }
+        this.collection = collection;
+        this.lock = newLock == null ? new ReentrantReadWriteLock() : newLock;
+    }
+
+    public ReadWriteLock getLock()
+    {
+        return this.lock;
+    }
+
+
+    protected void acquireWriteLock()
+    {
+        this.getLock().writeLock().lock();
+    }
+
+    protected void unlockWriteLock()
+    {
+        this.getLock().writeLock().unlock();
+    }
+
+    protected void acquireReadLock()
+    {
+        this.getLock().readLock().lock();
+    }
+
+    protected void unlockReadLock()
+    {
+        this.getLock().readLock().unlock();
+    }
+
+
+    protected Mutable<name>Collection get<name>Collection()
+    {
+        return this.collection;
+    }
+
+    public int size()
+    {
+        <readLock()>
+            return this.collection.size();
+        <readUnlock()>
+    }
+
+    public boolean isEmpty()
+    {
+        <readLock()>
+            return this.collection.isEmpty();
+        <readUnlock()>
+    }
+
+    public boolean notEmpty()
+    {
+        <readLock()>
+            return this.collection.notEmpty();
+        <readUnlock()>
+    }
```

Comment:
Since PrimitiveIterable supports size(), isEmpty(), and notEmpty(), these methods could be pulled up as well. Depending on how many methods we find that can be pulled up, it could make decent impact on lines of generated code and jar size.